`@types/node` not recognized

```ts
{
  packageManager: 'pnpm',
  scriptModule: false,
  test: false,
  entryPoints: ['./mod.ts'],
  outDir: './npm',
  shims: {},
  compilerOptions: {
    target: "ES2021",
  },
  package: {
    name: '@itchatapp/migrations',
    version: Deno.args[0].replace(/[A-Z]+/gi, ''),
    description: 'migrations module for itchat',
    license: 'MIT',
    devDependencies: {
      "@types/node": "16.x",
    },
    repository: {
      type: 'git',
      url: 'git+https://github.com/itchatapp/migrations.git',
    },
    bugs: {
      url: 'https://github.com/itchatapp/migrations/issues',
    },
    engines: {
      node: '>=16.0.0',
    },
    files: [
      'esm/*',
      'types/*',
    ],
  },
}
```
<details>
<summary>Logs:</summary>

```
[dnt] Transforming...
[dnt] Running pnpm install...
Packages: +1
+
Packages are cloned from the content-addressable store to the virtual store.
  Content-addressable store is at: /home/abdulrahman/.local/share/pnpm/store/v3
  Virtual store is at:             node_modules/.pnpm
Progress: resolved 1, reused 1, downloaded 0, added 1, done

devDependencies:
+ @types/node 16.11.37 (17.0.38 is available)
[dnt] Building project...
[dnt] Type checking...
npm/src/src/lib.ts:1:22 - error TS2307: Cannot find module 'path' or its corresponding type declarations.

1 import { join } from 'path';
                       ~~~~~~
npm/src/src/lib.ts:2:16 - error TS2307: Cannot find module 'fs' or its corresponding type declarations.

2 import fs from 'fs';
                 ~~~~
npm/src/src/lib.ts:18:14 - error TS7006: Parameter 'x' implicitly has an 'any' type.

18     .filter((x) =>
                ~
npm/src/src/lib.ts:22:11 - error TS7006: Parameter 'x' implicitly has an 'any' type.

22     .map((x) => ({
             ~

error: Uncaught (in promise) Error: Had 4 diagnostics.
      throw new Error(`Had ${diagnostics.length} diagnostics.`);
```

</details>

version: 0.24.0 (I also tried 0.23.0 and 0.22.0 all of them the same result)

--- a/mod.ts
+++ b/mod.ts
@@ -275,15 +275,26 @@ export async function build(options: BuildOptions): Promise<void> {
     }
   }
 
-  let program = project.createProgram();
+  // When creating the program and type checking, we need to ensure that
+  // the cwd is the directory that contains the node_modules directory
+  // so that TypeScript will read it and resolve any @types/ packages.
+  // This is done in `getAutomaticTypeDirectiveNames` of TypeScript's code.
+  const originalDir = Deno.cwd();
+  let program: ts.Program;
+  Deno.chdir(options.outDir);
+  try {
+    program = project.createProgram();
 
-  if (options.typeCheck) {
-    log("Type checking...");
-    const diagnostics = ts.getPreEmitDiagnostics(program);
-    if (diagnostics.length > 0) {
-      outputDiagnostics(diagnostics);
-      throw new Error(`Had ${diagnostics.length} diagnostics.`);
+    if (options.typeCheck) {
+      log("Type checking...");
+      const diagnostics = ts.getPreEmitDiagnostics(program);
+      if (diagnostics.length > 0) {
+        outputDiagnostics(diagnostics);
+        throw new Error(`Had ${diagnostics.length} diagnostics.`);
+      }
     }
+  } finally {
+    Deno.chdir(originalDir);
   }
 
   // emit only the .d.ts files

--- a/tests/integration.test.ts
+++ b/tests/integration.test.ts
@@ -700,6 +700,23 @@ Deno.test("should build undici project", async () => {
   });
 });
 
+Deno.test("should build and type check node types project", async () => {
+  await runTest("node_types_project", {
+    scriptModule: false,
+    test: false,
+    entryPoints: ["main.ts"],
+    outDir: "./npm",
+    shims: {},
+    package: {
+      name: "node_types",
+      version: "0.0.0",
+      devDependencies: {
+        "@types/node": versions.nodeTypes,
+      },
+    },
+  });
+});
+
 export interface Output {
   packageJson: any;
   npmIgnore: string;
@@ -716,6 +733,7 @@ async function runTest(
     | "polyfill_project"
     | "polyfill_array_find_last_project"
     | "module_mappings_project"
+    | "node_types_project"
     | "undici_project"
     | "shim_project"
     | "test_project"

--- a/tests/node_types_project/main.ts
+++ b/tests/node_types_project/main.ts
@@ -1,0 +1,5 @@
+import { join } from "https://deno.land/std@0.142.0/node/path.ts";
+import fs from "https://deno.land/std@0.142.0/node/fs.ts";
+
+console.log(join("test", "other"));
+fs.writeFileSync("test.txt", "test");

